src/context/ProfileContext.tsx

```diff
@@ -57,87 +67,128 @@ export const useProfileContext = () => {
 export const ProfileProvider = ({ children }: { children: ReactNode }) => {
   const profileQuery = useProfiles();
   const [defaultProfileSelected, setDefaultProfileSelected] =
-    useState<Profile | null>(null);
+    useState<ExtendedProfile | null>(null);
   const { data: profiles } = profileQuery;
   const { data: lastProfile } = useLastProfile();
   const [localProfileIndex, setLocalProfileIndex] = useState<number>(0);
-  const [localProfile, setLocalProfile] = useState<Profile | null>(null);
+  const [localProfile, setLocalProfile] = useState<ExtendedProfile | null>(
+    null
+  );
   const [localHoverState, setLocalHoverState] = useState<boolean>(false);
   const [profileIdToFind, setProfileIdToFind] = useState<string | null>(null);
   const [profileEvent, setProfileEvent] = useState<ProfileUpdate | null>(null);
   const [profileStarting, setProfileStarting] = useState(false);
   const [settingsIndex, setSettingsIndex] = useState(0);
   const [settingsProfile, setSettingsProfile] = useState<
-    (Profile & { settings: IPresetSetting[] }) | null
+    (ExtendedProfile & { settings: IPresetSetting[] }) | null
   >(null);
 
+  const [hasJustHandledProfileEvent, setHasJustHandledProfileEvent] =
+    useState(false);
+
+  const mergedProfiles = useMemo<ExtendedProfile[]>(() => {
+    if (!profiles) return [];
+
+    const last = lastProfile?.profile;
+    if (!last) return profiles;
+
+    const existingIndex = profiles.findIndex((p) => p.id === last.id);
+    const existing = existingIndex !== -1 ? profiles[existingIndex] : null;
+
+    const profilesExtended = profiles.map((p) => ({
+      ...p,
+      isLast: false,
+      temporary: false
+    }));
+
+    //It exists in profiles and is identical.
+    if (existing && deepEqual(existing, last)) {
+      profilesExtended[existingIndex] = {
+        ...profilesExtended[existingIndex],
+        isLast: true
+      };
+      return profilesExtended;
+    }
+
+    //Add to the end (different content or does not exist)
```

Comment:
I love how clean this is compared to the redux implementation from 12 months ago. Nice!